.maintain/gitlab/check_companion_build.sh

```diff
@@ -0,0 +1,123 @@
+DEPS=("${@:4}")
+SUBSTRATE_DIR="$(pwd)"
+
+boldprint () { printf "|\n| \033[1m%s\033[0m\n|\n" "${@}"; }
+boldcat () { printf "|\n"; while read -r l; do printf "| \033[1m%s\033[0m\n" "${l}"; done; printf "|\n" ; }
+
+boldcat <<-EOT
+
+
+check_${REPO}_companion_build
+==============================
+
+this job checks if there is a string in the description of the pr like
+
+$REPO companion: $ORGANISATION/$REPO#567
+
+
+it will then run cargo check from this ${REPO}'s branch with substrate code
+from this pull request. otherwise, it will uses master instead
+
+
+EOT
+
+# Set the user name and email to make merging work
+git config --global user.name 'CI system'
+git config --global user.email '<>'
+
+# Merge master into our branch before building to make sure we don't miss
+# any commits that are required.
+git fetch --depth 100 origin
+git merge origin/master
+
+# Clone the current master branch into a local directory
+# NOTE: we need to pull enough commits to be able to find a common
+# ancestor for successfully performing merges below.
+git clone --depth 20 "https://github.com/${ORGANISATION}/${REPO}.git"
+
+cd "$REPO"
+
+# either it's a pull request then check for a companion otherwise use
+# master
+# shellcheck disable=SC2003
+if expr match "${CI_COMMIT_REF_NAME}" '^[0-9]\+$' >/dev/null
+then
+  boldprint "this is pull request no ${CI_COMMIT_REF_NAME}"
+  pr_companion="$(get_companion "paritytech/substrate" "$CI_COMMIT_REF_NAME" "$ORGANISATION/$REPO")"
+  if [ "$pr_companion" ]
+  then
+    boldprint "companion pr specified/detected: #${pr_companion}"
+    git fetch origin "refs/pull/${pr_companion}/head:pr/${pr_companion}"
+    git checkout "pr/${pr_companion}"
+    git merge origin/master
+  else
+    boldprint "no companion branch found - building ${REPO}:master"
+  fi
+
+  # If this repo has any additional dependencies, we should check whether they
+  # are mentioned as companions as well. If they are, we patch this repo to
+  # use that companion build as well. See example at top of this script
+  # Note: Will only work with repos supported by diener
+  declare -A diener_commands
+  diener_commands=()
+  diener_commands["paritytech/polkadot"]='--polkadot'
+  diener_commands["paritytech/substrate"]='--substrate'
+  diener_commands["paritytech/grandpa-bridge-gadget"]='--beefy'
+
+  for dep in "${DEPS[@]}"; do
+    dep_companion="$(get_companion "paritytech/substrate" "$CI_COMMIT_REF_NAME" "$dep")"
+    if [ "$dep_companion" ]; then
+      echo "Companion PR found for $dep, need to patch $REPO to use that"
+      git clone --depth 20 "https://github.com/$dep.git" "$dep"
+      git -C "$dep" fetch origin "refs/pull/${dep_companion}/head:pr/${dep_companion}"
+      git -C "$dep" checkout "pr/${dep_companion}"
+      git -C "$dep" merge origin/master
+      # Tell this dependency to use the version of substrate in this PR
+      diener patch --crates-to-patch "$SUBSTRATE_DIR" --substrate --path "$dep/Cargo.toml"
```

Comment:
I believe `SUBSTRATE_DIR` and `--substrate` will be reworked once this script is made more general (?)

Reply:
On another note, I think "$SUBSTRATE_DIR" is wrong here since you want to use the `pr/${pr_companion}` branch instead